cloudformation/stacks/AutoSpotting/template.json

```diff
@@ -160,6 +166,7 @@
                 "ec2:DescribeSpotInstanceRequests",
                 "ec2:DescribeSpotPriceHistory",
                 "ec2:RequestSpotInstances",
+                "ec2:CancelSpotInstanceRequests",
```

Comment:
Have you considered combining the `ValidUntil` parameter of [ec2.RequestSpotInstancesInput](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#RequestSpotInstancesInput) instead of handling all this ourselves?

This should considerably simplify the below logic, the spot instances would be launched and their request would be cleaned up automatically by AWS after it expires. Any launched instances would be kept running after the spot instance requests have been cleaned up.

Reply:
Hi there,

The ValidUntil, is a good point, but it won't address the other issue.  We need to look at the status of the spot instance request to see if it is in the holding state.  

Currently we wait on the spot instance request regardless, and this times out the lambda, and produces an invocation error (timed out).  From experience, I've seen requests in holding be there for hours.  This results in "errors" for hours.  Errors should really be an exceptional case, not a normal; as an error means it should be alerting someone to look into the issue.

Therefore for me, we still need to check the status of the request and not wait on it (just check it on the next lambda run).  

The ec2:CancelSpotInstanceRequests was already required, and missed, the following was pre-existing:

- https://github.com/cristim/autospotting/blob/master/core/autoscaling.go#L595

Reply:
The `validUntil` would be set to the maximum amount of time you want to allow instances in the holding state, for example five minutes.

If we didn't manage to launch any instances before this expires, the request is immediately cancelled by AWS without any action required on our side.

If the instance was launched, it would be kept running but the spot request would be cancelled nevertheless. I think this also has some potential to improve the throttling situation which happens quite often on large installations (50+ groups handled in parallel).

But we need to pay attention not to get any orphaned instances outside the groups, if their spot requests have since been cancelled. One way to do this is to also tag the instance with the group it needs to be added to, instead of just tagging the SIR

Reply:
but the ValidUntil is going to effect all spot instance requests that are made.  You are going to end up with a lot of spot instances, and cancelled spot requests.  Which is going to confuse users.  i.e. if I have a running spot instances then I expect an active spot instance request.

Reply:
That's not the case even today, the SIRs are cancelled after a week (the default value for `ValidUntil`) even if their instances keep running for longer time.

So as far as I can tell the SIR is meant to be temporary, only useful until an instance can be launched from it, then it can be safely discarded.

The source of truth should always be the running instances.

Reply:
Ah ok, we are using it in our Dev environments that shutdown daily, so nothing lasts a week.

So.  Given this, what do we wish to do:

- when a spot request is seen to be in holding set the valid until (can you set this on a existing request - or does it have to be done when the request is made?)

- can we tag the instance with the spot request it was created from?

The tagging of instance for which no spot request exists (is cancelled due to the valid until), sounds like quite a bit of work

Reply:
I say this is a lot of work, as the code base (core/autoscaling.go) is all about iterating open or active SIRs

Reply:
I don't see the ability to update a spot request after creation (only fleets, not the same thing).

I don't see valid-until working without a lot of work.

Reply:
Ok.  Do we wish to change:

-max_time_spot_request_can_be_holding=0:

To:

-seconds_spot_request_valid_for

And then change the spot request creation to use this to create the specified valid-until on the spot request?

----
I'm very unsure about the following:

```But we need to pay attention not to get any orphaned instances outside the groups, if their spot requests have since been cancelled. One way to do this is to tag the instance with the group it needs to be added to, instead of tagging the SIR. The instances are described already across the entire fleet, so we don't need to do any additional API calls, just to verify if they are tagged for the current group.

Unfortunately we need to wait for the SIR to launch instances because we need to make sure they have been tagged, because the SIR won't tag the instances for us and we have to do it as soon as they were launched.```

Reply:
I think the logic would have to enforce that SIRs are temporary and never live longer than a Lambda function run.

- We just launch the SIR and wait for the instance to come up using the waiter we already have in place (only so we can tag the instance), and clean up the SIR as soon as their instance was launched. We can also set the SIRs to be cancelled automatically after less than the 5 minutes Lambda timeout, just in case we have a crash so we don't have SIRs lying around.
- We remove all the logic of tracking and handling open SIRs between multiple function runs, as well as the tracking of intermediate SIR states(handled by the waiter anyway within the run) and also remove the cancelling logic you just introduced. This should simplify things a lot.
- I think it makes little sense to have the `-seconds_spot_request_valid_for` configurable, except for testing purposes. Most of the time a ~4min default should be fine.
- We tag the instance with the group it needs to be added to, not with the SIR. The next run will see the spot instances that have to be added to the current group and do the swapping.

This will be some work, but mainly cleaning up and simplifying the logic, I expect it to remove a few hundreds lines of code which is always great.

Reply:
```We just launch the SIR and wait for the instance to come up using the waiter we already have in place (only so we can tag the instance)```

Waiting for instances on an open, but in holding by Amazon, will cause the lambda to hang and time out (i.e. an Error).   We would still have to check for "open" but in holding, so we don't unnecessarily wait on SIR in holding

----

```I think it makes little sense to have the -seconds_spot_request_valid_for configurable, except for testing purposes. Most of the time a ~4min default should be fine.```

If we set the default for 4 mins, then the logic for finding SIR and instances all has to change (`havingReadyToAttachSpotInstance()` in autoscaling.go and logic around it); and I would say that having it configurable would turn out to be something required.

----

All of the above is really a different requirement:

- don't use SIR as the means for adding instances to ASGs

Yup, it can (if we are careful) address the don't waiting on stuff that Amazon puts in holding.  

This pull is specifically about improving the current situation.  It waits unnecessarily on SIR's in holding.  The default operation is as it was previously (never cancel a SIR that is in holding).  The change is to improve it by not waiting on the held SIR.  

Given this. 

If I was to update the pull request to not include the command line option for

- max_time_spot_request_can_be_holding

and update it so that we don't cancel the spot and don't compare time it has been open. So that nothing leaks into the public interface of the lambda (all behind the scenes changes)

Then we get:

- missing ec2:CancelSpotInstanceRequests fix
- nil pointer exception fix by checking for errors on the waitForAndTagSpotInstance(), and not derefencing a nil id: (https://github.com/cristim/autospotting/blob/master/core/autoscaling.go#L584)
- not waiting on holding requests.

Then I think as a separate pull request I can potentially look into the more complex task/requirement of:

- don't use SIR as the means for adding instances to ASGs


Does that make sense?

Reply:
Sure, let's do this!

@xlr-8 what are your thoughts on this?

Reply:
Just pondering if I'm good to do the baby steps?  I might get a couple of free hours tmr.

Reply:
By all means!

Feel free to get in touch on gitter if you have any questions or concerns.

Huge thanks for volunteering to do this, it'll be a great improvement.